benchmarks/canny_edge_detector/procedural_c#/Program.cs

```diff
@@ -0,0 +1,307 @@
+
+		public static int[,] Convolve(int[,] image, double[,] kernel)
+		{
+			var width = image.GetLength(0);
+			var height = image.GetLength(1);
+			var test = new int[width, height];
+			var halfKernel = kernel.GetLength(0) / 2;
+			for (int x = 0; x < width; x++)
+			{
+				for (int y = 0; y < height; y++)
+				{
+					var sum = 0.0;
+					for (int kx = -halfKernel; kx <= halfKernel; kx++)
+					{
+						for (int ky = -halfKernel; ky <= halfKernel; ky++)
+						{
+							var posX = x + kx;
+							var posY = y + ky;
+							// not edges
+							if (!(posX <= 0 || posX >= width - 1 || posY <= 0 || posY >= height - 1))
+							{
+								sum += (image[posX, posY] * (kernel[kx + halfKernel, ky + halfKernel]));
+							}
+						}
+					}
+					if (sum > 255.0)
+					{
+						sum = 255.0;
+					}
+					else if (sum < 0.0)
+					{
+						sum = 0.0;
+					}
+
+					test[x, y] = (int)sum;
+				}
+			}
+			return test;
+		}
+
+		public static int hyp(int num1, int num2)
+		{
+			var hyp = (Math.Sqrt((num1 * num1) + (num2 * num2)));
+			return Math.Min(255, (int)hyp);
+		}
+
+		public static int[,] hypot(int[,] image1, int[,] image2)
+		{
+			var width = image1.GetLength(0);
+			var height = image1.GetLength(1);
+			var result = new int[width, height];
+
+			for (int x = 0; x < width; x++)
+			{
+				for (int y = 0; y < height; y++)
+				{
+					var color1 = image1[x, y];
+					var color2 = image2[x, y];
+
+					var hypColor = hyp(color1, color2);
+					result[x, y] = hypColor;
+				}
+			}
+			return result;
+		}
+
+		public static double[,] arctan(int[,] image1, int[,] image2)
+		{
+			var width = image1.GetLength(0);
+			var height = image1.GetLength(1);
+			var result = new double[width, height];
+
+			for (int x = 0; x < width; x++)
+			{
+				for (int y = 0; y < height; y++)
+				{
+					var color1 = image1[x, y];
+					var color2 = image2[x, y];
+					result[x, y] = Math.Atan2(color1, color2);
+				}
+			}
+			return result;
+		}
+
+		public static (int[,], int[,]) computeIntensity(int[,] image)
+		{
+			var Ix = Convolve(image, kernelHor);
+			var Iy = Convolve(image, kernelVer);
+
+			var g = hypot(Ix, Iy);
+
+			var theta = arctan(Iy, Ix);
+			var thetaWidth = theta.GetLength(0);
+			var thetaHeight = theta.GetLength(1);
+			var thetaQ = new int[thetaWidth, thetaHeight];
+			for (int i = 0; i < thetaWidth; i++)
+			{
+				for (int j = 0; j < thetaHeight; j++)
+				{
+					var num = ((int)(Math.Round(theta[i, j] * (5.0 / Math.PI))) + 5) % 5;
+					thetaQ[i, j] = num;
+				}
+			}
+
+			return (g, thetaQ);
+		}
+
+		public static int[,] nonMaxSuppresion(int[,] image, int[,] theta)
+		{
+			// Non-maximum suppression
+			var width = image.GetLength(0);
+			var height = image.GetLength(1);
+			var gradSup = image;
+
+			for (int r = 0; r < width; r++)
+			{
+				for (int c = 0; c < height; c++)
+				{
+					//Suppress pixels at the image edge
+					if (r == 0 || r == width - 1 || c == 0 || c == height - 1)
+					{
+						gradSup[r, c] = 0;
+					}
```

Comment:
Hvis du kører fra r=1 til width-1 og c=1 til height-1 så slipper du for den her if